html/entities.ts

```diff
@@ -103,3 +115,73 @@ function codePointStrToChar(codePointStr: string, radix: number) {
 
   return codePoint > MAX_CODE_POINT ? "�" : String.fromCodePoint(codePoint);
 }
+
+/**
+ * Normalization form to use for escaping. See {@linkcode normalize} for examples.
+ *
+ * - `readability`: Optimize for human readability and file size.
+ * - `compatibility`: Optimize for compatibility across boundaries that lack
+ *    full Unicode support, are unaware of encoding, or fail to respect
+ *    encoding.
+ */
+export type NormalizationForm =
+  | "readability"
+  | "compatibility";
+
+export type NormalizationOptions = { form: NormalizationForm };
+const defaultNormalizationOptions: NormalizationOptions = {
+  form: "readability",
+};
+
+/**
+ * Normalize HTML or XML entities in a string of markup.
+ *
+ * @example
+ * ```ts
+ * import { normalize } from "https://deno.land/std@$STD_VERSION/html/entities.ts";
+ *
+ * normalize("&#x3e;"); // "&gt;"
+ * normalize("&apos;"); // "&#39;"
+ * normalize("&#x4e24;&#x53ea;&#x5c0f;&#x871c;&#x8702;"); // "两只小蜜蜂"
+ *
+ * // other markup is left untouched
+ * normalize("<p class='foo'>&#x1f308;</p>") // "<p class='foo'>🌈</p>"
+ *
+ * // specifying a `form` option (default is `readability`):
+ * normalize("两只小蜜蜂", { form: "readability" }); // "两只小蜜蜂"
+ * normalize("两只小蜜蜂", { form: "compatibility" }); // "&#x4e24;&#x53ea;&#x5c0f;&#x871c;&#x8702;"
+ * ```
+ */
+export function normalize(
+  str: string,
+  options?: Partial<NormalizationOptions>,
+) {
+  const { form } = { ...defaultNormalizationOptions, ...options };
+  return str
+    .split(/([<>'"]+)/)
+    .map((segment, i) => i % 2 ? segment : escape(unescape(segment), { form }))
+    .join("");
+}
+
+/**
+ * See https://en.wikipedia.org/wiki/Valid_characters_in_XML#Non-restricted_characters
+ */
+function escapeXmlRestricted(str: string) {
+  return str.replaceAll(
+    // deno-lint-ignore no-control-regex
+    /[^\x09\x0a\x0d\x20-\x7e\x85\xa0-\ud7ff\ue000-\ufdcf\ufdf0-\ufffd\u{10000}-\u{1fffd}\u{20000}-\u{2fffd}\u{30000}-\u{3fffd}\u{40000}-\u{4fffd}\u{50000}-\u{5fffd}\u{60000}-\u{6fffd}\u{70000}-\u{7fffd}\u{80000}-\u{8fffd}\u{90000}-\u{9fffd}\u{a0000}-\u{afffd}\u{b0000}-\u{bfffd}\u{c0000}-\u{cfffd}\u{d0000}-\u{dfffd}\u{e0000}-\u{efffd}\u{f0000}-\u{ffffd}\u{100000}-\u{10fffd}]+/gu,
```

Comment:
Why do we start escaping these chars by default?

Also does HTML have the same restricted chars concept as XML?

Reply:
Upon reflection, this PR needs some more thought. I'd initially thought a sort of "baseline-compatible with both HTML and XML" would be a sensible default, given that these characters are rare in practice yet could cause problems in XML. But it turns out that entities for certain C1 control-character codepoints have [different semantics in HTML](https://html.spec.whatwg.org/multipage/parsing.html#numeric-character-reference-end-state) than XML — for example, `&#x80;`:

```js
['application/xml', 'text/html'].map((contentType) => {
    const { literal, entity } = JSON.parse(
        new DOMParser().parseFromString(
            '<div>{"literal": "\x80", "entity": "&#x80;"}</div>',
            contentType,
        ).querySelector('div').textContent,
    )
    return { contentType, literal, entity }
})

// { "contentType": "application/xml", "literal": "\x80", "entity": "\x80" }
// { "contentType": "text/html", "literal": "\x80", "entity": "€" }
```

Also, I'm not sure converting those characters to entities is the right approach. Probably simply stripping them out would be more sensible in most cases.